segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,453 @@
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Storage adapter for file system based Tier2.
+ *
+ * Each segment is represented as a single file on the underlying storage. As the writes at an offset for a POSIX
+ * file is idempotent, there is no need for locking when a container fails over to another host.
+ * As Pravega does not modify data in Tier2 once written even a contention in writing will cause same data to be
+ * written at the same offset till the time the original host gives up ownership.
+ */
+
+@Slf4j
+public class FSStorage implements Storage {
+
+    //region members
+
+    private final FSStorageConfig config;
+    private final ExecutorService executor;
+
+    //endregion
+
+    //region constructor
+
+    public FSStorage(FSStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+    @Override
+    public void initialize(long containerEpoch) {
+
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        final CompletableFuture<SegmentHandle> retVal = new CompletableFuture<>();
+
+        executor.execute( () -> {
+            syncOpenRead(streamSegmentName, retVal);
+        });
+
+        return retVal;
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        final CompletableFuture<Integer> retVal = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncRead(handle, offset, buffer, bufferOffset, length, timeout, retVal);
+        });
+
+        return retVal;
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        final CompletableFuture<SegmentProperties> retVal = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncGetStreamSegmentInfo(streamSegmentName, timeout, retVal);
+        });
+
+        return retVal;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        final CompletableFuture<Boolean> retFuture = new CompletableFuture<>();
+
+        executor.execute(() -> {
+            syncExists(streamSegmentName, timeout, retFuture);
+        });
+
+        return retFuture;
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
```

Comment:
As discussed in the comments earlier, we do not need either distributed or in proc synchronization.